.changeset/friendly-hairs-happen.md

```diff
@@ -0,0 +1,62 @@
+---
+'@backstage/plugin-catalog-backend-module-aws': patch
+---
+
+Add a new provider `AwsS3EntityProvider` as replacement for `AwsS3DiscoveryProcessor`.
+
+In order to migrate from the **deprecated** `AwsS3DiscoveryProcessor` you need to apply
+the following changes:
+
+**Before:**
+
+```yaml
+# app-config.yaml
+
+catalog:
+  locations:
+    - type: s3-discovery
+      target: https://sample-bucket.s3.us-east-2.amazonaws.com/prefix/
+```
+
+```ts
+/* packages/backend/src/plugins/catalog.ts */
+
+import { AwsS3DiscoveryProcessor } from '@backstage/plugin-catalog-backend-module-aws';
+
+const builder = await CatalogBuilder.create(env);
+/** ... other processors ... */
+builder.addProcessor(new AwsS3DiscoveryProcessor(env.reader));
+```
+
+**After:**
+
+```yaml
+# app-config.yaml
+
+catalog:
```

Comment:
Hmm, I'm not sure about adding a new entityProvider and placing the configuration under `catalog.processors.awsS3.providers`. Perhaps we can do better? @backstage/maintainers suggestions?

Reply:
I tried to follow the example of existing providers like ldapOrg or microsoftGraphOrg. But I get what you mean. Open for suggestions :)

Reply:
maybe an alternative which could be applied here and for the upcoming processor to provider conversion as of #10183 

```yaml
catalog:
  providers:
    awsS3:
      - bucket: bucket-name
        # [...]
      - bucket: another-bucket-name
        # [...]
```

I guess it would be great to migrate existing providers to it, too, then.

Reply:
values like bucket names are usually environment-specific. Doing it "hardcoded"/in code would require me still to put it in my own config and load it from there instead.

And if I think about existing providers, this approach might be tricky for them.

But I'm generally more in favor of using the config file.

Reply:
seems like we only need alignment/a decision here to move on :)

Reply:
I will opt-in for `catalog.providers.awsS3` for now. Let me know what you think

Reply:
btw: we also cases like `catalog.processors.awsOrganization.provider` used for the **processor** `AwsOrganizationCloudAccountProcessor`

Reply:
The main thing I'm worried about is supporting multiple instances of the provider, which can get tricky depending on how we structure the config. Both for this specific provider and as a general pattern. Totally agree that if this is often environment specific then there should be some built-in way to have that information in config. Ideas welcome 😅

Reply:
the current setup assumes multiple instances by default (config array) and the factory returns an array of providers as well

Reply:
Of course, one provider instance could handle all buckets (+ prefixes), too, and iterate over them/the config array. But multiple instances can help to spread the load is what I've considered here for this specific provider.

There might be cases where a single instance of a provider is enough. As a general pattern, supporting multiple seems still better. But this does not need to be limited, too. `catalog.providers.{name}` as config prefix could be enough unification/requirement. We could expect providers to always support multiple instances (the value is a config array with the config per instance) or just leave it open how it is structured (single, multiple, multiple with some basic config, etc.).

Reply:
To clarify I'm thinking multiple providers of the same type but with different options and configuration. Think for example 3 different buckets with different scheduling configuration, or custom transformation functions. It's not something that existing providers really support and I don't think we have to support it in this PR, but I would like to find a pattern that works well with that so curious if you have any thoughts.

Reply:
With a shared base interface including scheduling options some of this could even be moved up the chain a bit and force some unification -- if desired :)

Reply:
Hm. I guess it can be argued that the config here is an `AwsS3DiscoveryProviderFactory` concern, not an `AwsS3EntityProvider` concern. So it would actually be up to the adopter whether they want to / need to make a non-blessed / esoteric setup where they manually instantiate providers, or whether they leverage the fast-lane factory helper. So ... from THAT angle maybe this is entirely OK.

Reply:
What do you think about adding
```
schedule:
  frequency: PT30M # or cron
  timeout: PT3M
  initialDelay: PT5M # optional
```
to the config read by the factory?

Reply:
Hold off on that. Good idea, but if we'd go that path, we would want it to be 100% powered by backend-tasks and having some form of fromConfig in there to read that config in a centralised way. Some other day, let's get this one in first

Reply:
alright, I will skip this for now

Reply:
I'd say that the transforms are only specific to the `MicrosoftGraphOrgEntityProvider` right now, expecting that pattern to show up in more places tbh.

Potentially something like this?

```yaml
catalog:
  providers:
    awsS3:
      default:
        bucket: bucket-name
```

Along with 

```ts
AwsS3DiscoveryProviderFactory.fromConfig(env.config, { logger: env.logger });
// OR
AwsS3DiscoveryProviderFactory.fromConfig(env.config, { logger: env.logger, id: 'default' });
```

With multiple providers just being more keys with different IDs.

Then another option that avoids having to always add the extra config level:

```yaml
catalog:
  providers:
    awsS3:
      # Not an array, implicit 'default'
      bucket: bucket-name
```

```yaml
catalog:
  providers:
    awsS3:
      # Array format, id is required
    - id: 'default'
      bucket: bucket-name
    - id: 'other'
      bucket: bucket-name
```

Again, just bouncing some ideas :grin: Although see the separate comment about wanting stable IDs, which ends up being a bit related to this

Reply:
@Rugvip I guess there is the question whether we want to support the implicit ID option (single provider case) or always assume multiple.

For this provider, multiple seems fine to me, but as a general pattern, maybe supporting both is desired and then doing it here would be reasonable (and not hard anyways).

Probably, I would go for first example using named configs instead of the array.

Reply:
I use the following approach now
```
catalog:
  providers:
    awsS3:
      providerId:
        bucket: bucket-name
```